metals/src/main/scala/scala/meta/internal/metals/CallHierarchyProvider.scala

```diff
@@ -0,0 +1,376 @@
+              source,
+              doc,
+              result.occurrence.get,
+              definition.pos.toLSP
+            )
+          } yield chi
+        }
+      case None =>
+        Nil
+    }
+  }
+
+  private def findIncomingCalls(
+      source: AbsolutePath,
+      doc: TextDocument,
+      root: Tree,
+      info: CallHierarchyItemInfo
+  ): List[(SymbolOccurrence, l.Range, List[l.Range])] = {
+    def search(
+        tree: Tree,
+        parent: Option[Name],
+        parentRange: Option[l.Range]
+    ): List[(SymbolOccurrence, l.Range, l.Range)] = tree match {
+      case name: Name
+          if (parent.isDefined && definition
+            .positionOccurrences(source, name.pos.toLSP.getEnd, doc)
+            .flatMap(_.occurrence)
+            .exists(occ =>
+              occ.symbol == info.symbol && occ.role == Role.REFERENCE
+            )) ||
+            (name.parent.map(_.is[Term.Apply]).getOrElse(false) && definition
+              .positionOccurrences(source, name.pos.toLSP.getEnd, doc)
+              .flatMap(_.occurrence)
+              .exists(occ =>
+                info.symbol.startsWith(occ.symbol.dropRight(1))
+              )) =>
+        definition
+          .positionOccurrences(source, parent.get.pos.toLSP.getStart, doc)
+          .flatMap(_.occurrence)
+          .map(occurence => (occurence, name.pos.toLSP, parentRange.get))
+      case method: Defn.Def =>
+        method.children.flatMap { child =>
+          search(child, Some(method.name), Some(method.pos.toLSP))
+        }
+      case constructor: Ctor.Secondary =>
+        constructor.children.flatMap { child =>
+          search(
+            child,
+            Some(constructor.name: Name),
+            Some(constructor.pos.toLSP)
+          )
+        }
+      case function: Term.Function =>
+        function.children.flatMap { child =>
+          val name = function.parent.flatMap {
+            case v: Defn.Val =>
+              v.pats match {
+                case (pat: Pat.Var) :: Nil => Some(pat.name)
+                case _ => None
+              }
+            case v: Defn.Var =>
+              v.pats match {
+                case (pat: Pat.Var) :: Nil => Some(pat.name)
+                case _ => None
+              }
+            case _ => None
+          }
+          search(child, name, function.parent.map(_.pos.toLSP))
+        }
+      case other =>
+        other.children.flatMap { child =>
+          search(child, parent, parentRange)
+        }
+    }
+    search(root, None, None)
+      .groupBy(_._1)
+      .map { case (k, v) => (k, v.head._3, v.map(_._2)) }
+      .toList
+  }
+
+  def incomingCalls(
+      params: CallHierarchyIncomingCallsParams
+  ): List[CallHierarchyIncomingCall] = {
+    val source = params.getItem.getUri.toAbsolutePath
+
+    val info = params.getItem.getData.asInstanceOf[JsonObject]
+
+    semanticdbs.textDocument(source).documentIncludingStale match {
```

Comment:
You might need to use ReferenceProvider to find all the documents that contain the symbol we are interested in. And then just iterate over them.

Reply:
Do you think I should use the public `references method` and get all URIs ?

Reply:
We could, or we could add a new method if that would be easier to handle. That method could just give the potential files based on Bloom filters